midend/lib/Utils/DIPUtils.cpp

```diff
@@ -410,6 +411,185 @@ Value customTanVal(OpBuilder &builder, Location loc, Value angleVal) {
   return builder.create<arith::DivFOp>(loc, sinVal, cosVal);
 }
 
+SmallVector<Value, 6> getRotationMatrix(OpBuilder& builder, Location loc, Value centerX, Value centerY, Value angle, Value scale) {
+  Value alpha0 = builder.create<math::CosOp>(loc, angle);
+  Value alpha = builder.create<arith::MulFOp>(loc, alpha0, scale);
+  Value beta0 = builder.create<math::SinOp>(loc, angle);
+  Value beta = builder.create<arith::MulFOp>(loc, beta0, scale);
+  Value oneMinusAlpha = builder.create<arith::SubFOp>(loc, builder.create<arith::ConstantOp>(loc, builder.getF32FloatAttr((float)1.)), alpha);
+  SmallVector<Value, 6> mat;
+  Value m20 = builder.create<arith::MulFOp>(loc, oneMinusAlpha, centerX);
+  Value m21 = builder.create<arith::MulFOp>(loc, beta, centerY);
+  Value m50 = builder.create<arith::MulFOp>(loc, beta, centerX);
+  Value m51 = builder.create<arith::MulFOp>(loc, oneMinusAlpha, centerY);
+  Value m0 = alpha;
+  Value m1 = beta;
+  Value m2 = builder.create<arith::SubFOp>(loc, m20, m21);
+  Value m3 = builder.create<arith::NegFOp>(loc, beta);
+  Value m4 = alpha;
+  Value m5 = builder.create<arith::AddFOp>(loc, m50, m51);
+
+  return SmallVector<Value, 6>{m0, m1, m2, m3, m4, m5};
+}
+
+inline void inverseAffineMatrix(OpBuilder &builder, Location loc,
+                                SmallVector<Value, 6> &affineMatrix) {
+  Value c0F32 = builder.create<arith::ConstantOp>(
+      loc, builder.getF32FloatAttr((float).0));
+  Value m0pm4 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[0], affineMatrix[4]);
+  Value m1pm3 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[1], affineMatrix[3]);
+  Value D = builder.create<arith::SubFOp>(loc, m0pm4, m1pm3);
+  // TODO: can a float number compare with 0.0 directly?
```

Comment:
What exactly is the intent behind this comment?

Reply:
I don't know what will happen in mlir if I wrote a line of code: if(x == 0.0). If I needed to do a comparison of a floating point number to a zero using C, I would write if(fabs(x) < 1e-6). So I'm not sure if this **is** absolutely correct.

Reply:
> if(x == 0.0)

This will not generate comparison code. If you wish to generate IR that does comparisons at runtime, you need to use `cmp` ops. Lowering pass generates lowered IR at compile time which can then be used to form an executable.

Reply:
This comment is for line 456 and 457

Reply:
Then perhaps put this comment at that location? or make this intent more clear by mentioning this in the comment itself?

Reply:
Okay

Reply:
Is this resolved?